Frame.cc

```diff
@@ -1417,3 +1359,116 @@ bool Frame::GetLoaderSpectralData(int region_id, int stokes, const casacore::Arr
     // Get spectral data from loader (add image mutex for swizzled data)
     return _loader->GetRegionSpectralData(region_id, stokes, mask, origin, _image_mutex, results, progress);
 }
+
+bool Frame::CalculateMoments(int file_id, MomentProgressCallback progress_callback, const casacore::ImageRegion& image_region,
+    const CARTA::MomentRequest& moment_request, CARTA::MomentResponse& moment_response,
+    std::vector<carta::CollapseResult>& collapse_results) {
+    if (!_moment_generator) {
+        _moment_generator = std::make_unique<MomentGenerator>(GetFileName(), GetImage());
+    }
+    if (_moment_generator) {
+        std::unique_lock<std::mutex> ulock(_image_mutex); // Must lock the image while doing moment calculations
+        _moment_generator->CalculateMoments(
+            file_id, image_region, _spectral_axis, _stokes_axis, progress_callback, moment_request, moment_response, collapse_results);
+        ulock.unlock();
+    }
+
+    return !collapse_results.empty();
+}
+
+void Frame::StopMomentCalc() {
+    if (_moment_generator) {
+        _moment_generator->StopCalculation();
+    }
+}
+
+void Frame::SaveFile(const std::string& root_folder, const CARTA::SaveFile& save_file_msg, CARTA::SaveFileAck& save_file_ack) {
+    // Input file info
+    std::string in_file = GetFileName();
+    casacore::ImageInterface<float>* image = GetImage();
+
+    // Output file info
+    std::string output_filename(save_file_msg.output_file_name());
+    std::string directory(save_file_msg.output_file_directory());
+    CARTA::FileType output_file_type(save_file_msg.output_file_type());
+
+    // Set response message
+    int file_id(save_file_msg.file_id());
+    save_file_ack.set_file_id(file_id);
+    bool success(false);
+    casacore::String message;
+
+    // Get the full resolved name of the output image
+    std::string temp_path = root_folder + "/" + directory;
+    std::string abs_path = fs::absolute(fs::path(temp_path)).string();
+    output_filename = abs_path + "/" + output_filename;
```

Comment:
This is a minor issue, but it would be cleaner to [use the `/` operator on path objects](https://en.cppreference.com/w/cpp/filesystem/path/operator_slash) here instead of concatenating raw strings with a hardcoded separator.